File: src/main/java/net/katsstuff/danmakucore/entity/EntityFallingData.java
/*
 * This class was created by <Katrix>. It's distributed as
 * part of the DanmakuCore Mod. Get the Source Code in github:
 * https://github.com/Katrix-/DanmakuCore
 *
 * DanmakuCore is Open Source and distributed under the
 * the DanmakuCore license: https://github.com/Katrix-/DanmakuCore/blob/master/LICENSE.md
 */
package net.katsstuff.danmakucore.entity;

import java.util.UUID;

import javax.annotation.Nullable;

import net.katsstuff.danmakucore.CoreDataSerializers;
import net.katsstuff.danmakucore.data.Vector3;
import net.katsstuff.danmakucore.helper.NBTHelper;
import net.katsstuff.danmakucore.helper.TouhouHelper;
import net.minecraft.entity.Entity;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.init.SoundEvents;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.network.datasync.DataParameter;
import net.minecraft.network.datasync.DataSerializer;
import net.minecraft.network.datasync.EntityDataManager;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;

public class EntityFallingData extends Entity {

	public enum DataType {
		SCORE_GREEN,
		SCORE_BLUE,
		POWER,
		BIG_POWER,
		LIFE,
		BOMB
	}

	public static final DataSerializer<DataType> DATA_TYPE_SERIALIZER = new CoreDataSerializers.EnumSerializer<>(DataType.class);

	private static final DataParameter<DataType> DATA_TYPE = EntityDataManager.createKey(EntityFallingData.class, DATA_TYPE_SERIALIZER);

	@Nullable
	private Entity target;
	private Vector3 direction;
	private float amount;

	public EntityFallingData(World worldIn) {
		super(worldIn);
		setSize(0.5F, 0.5F);
	}

	public EntityFallingData(World world, DataType dataType, Vector3 pos, Vector3 direction, @Nullable Entity target, float amount) {
		this(world);
		setDataType(dataType);
		setPositionAndRotation(pos.x(), pos.y(), pos.z(), (float)direction.yaw(), (float)direction.pitch());
		this.target = target;
		this.direction = direction;
		this.amount = amount;
	}

	@Override
	protected void entityInit() {
		dataManager.register(DATA_TYPE, DataType.SCORE_GREEN);
	}

	@Override
	public void onUpdate() {
		super.onUpdate();

		if(!world.isRemote) {

			Vector3 motion;
			if(target != null) {
				motion = Vector3.directionToEntity(this, target);
			}
			else {
				motion = direction.multiply(0.25);
			}

			motionX = motion.x();
			motionY = motion.y();
			motionZ = motion.z();

			if(!world.isAirBlock(new BlockPos(posX + motionX, posY + motionY, posZ + motionZ))) {
				setDead();
				return;
			}

			setPosition(posX + motionX, posY + motionY, posZ + motionZ);
		}
	}

	@Override
	public void onCollideWithPlayer(EntityPlayer player) {
		if(!world.isRemote) {
			switch(getDataType()) {
				case SCORE_GREEN:
				case SCORE_BLUE:
					TouhouHelper.changeAndSyncPlayerData(data -> data.addScore((int)amount), player);
					break;
				case POWER:
				case BIG_POWER:
					TouhouHelper.changeAndSyncPlayerData(data -> data.addPower(amount), player);
					break;
				case LIFE:
					TouhouHelper.changeAndSyncPlayerData(data -> data.addLives((int)amount), player);
					break;
				case BOMB:
					TouhouHelper.changeAndSyncPlayerData(data -> data.addBombs((int)amount), player);
					break;
				default:
					break;
			}
			this.playSound(SoundEvents.ENTITY_ITEM_PICKUP, 0.75F, 1.3F);
			setDead();
		}
	}

	public DataType getDataType() {
		return dataManager.get(DATA_TYPE);
	}

	public void setDataType(DataType dataType) {
		dataManager.set(DATA_TYPE, dataType);
	}

	@Override
	protected void readEntityFromNBT(NBTTagCompound compound) {
		UUID target = compound.getUniqueId("target");
		if(target != null) {
			Entity targetEntity = world.getPlayerEntityByUUID(target);
			if(targetEntity == null) {
				NBTHelper.getEntityByUUID(target, world);
			}
			this.target = targetEntity;
		}

		direction = NBTHelper.getVector(compound, "direction");
		amount = compound.getFloat("amount");

		setDataType(DataType.class.getEnumConstants()[compound.getInteger("dataType")]);
	}

	@Override
	protected void writeEntityToNBT(NBTTagCompound compound) {
		if(target != null) {
			compound.setUniqueId("target", target.getUniqueID());
		}

		NBTHelper.setVector(compound, "direction", direction);
		compound.setFloat("amount", amount);
		compound.setInteger("dataType", getDataType().ordinal());
	}
}

File: src/main/scala/net/katsstuff/danmakucore/entity/EntityFallingData.scala
/*
 * This class was created by <Katrix>. It's distributed as
 * part of the DanmakuCore Mod. Get the Source Code in github:
 * https://github.com/Katrix-/DanmakuCore
 *
 * DanmakuCore is Open Source and distributed under the
 * the DanmakuCore license: https://github.com/Katrix-/DanmakuCore/blob/master/LICENSE.md
 */
package net.katsstuff.danmakucore.entity

import net.katsstuff.danmakucore.data.Vector3
import net.katsstuff.danmakucore.entity.EntityFallingData.DataType
import net.katsstuff.danmakucore.entity.EntityFallingData.DataType.{BigPower, Bomb, Life, Power, ScoreBlue, ScoreGreen}
import net.katsstuff.danmakucore.helper.NBTHelper
import net.katsstuff.danmakucore.lib.{LibEntityName, LibSounds}
import net.katsstuff.danmakucore.misc.LogicalSideOnly
import net.katsstuff.danmakucore.scalastuff.DanCoreImplicits._
import net.katsstuff.danmakucore.scalastuff.TouhouHelper
import net.minecraft.entity.Entity
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.network.datasync.{DataSerializer, DataSerializers, EntityDataManager}
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World
import net.minecraftforge.fml.relauncher.Side

/**
  * JAVA API
  *
  * Allows access to the different data types from Java.
  */
object FallingDataTypes {
  def scoreGreen: DataType = ScoreGreen
  def scoreBlue:  DataType = ScoreBlue
  def power:      DataType = Power
  def bigPower:   DataType = BigPower
  def life:       DataType = Life
  def bomb:       DataType = Bomb
}
object EntityFallingData {
  trait DataType
  object DataType {
    case object ScoreGreen extends DataType
    case object ScoreBlue  extends DataType
    case object Power      extends DataType
    case object BigPower   extends DataType
    case object Life       extends DataType
    case object Bomb       extends DataType

    def fromId(id: Byte): Option[DataType] = id match {
      case 0 => Some(ScoreGreen)
      case 1 => Some(ScoreBlue)
      case 2 => Some(Power)
      case 3 => Some(BigPower)
      case 4 => Some(Life)
      case 5 => Some(Bomb)
      case _ => None
    }

    def toId(obj: DataType): Byte = obj match {
      case ScoreGreen => 0
      case ScoreBlue  => 1
      case Power      => 2
      case BigPower   => 3
      case Life       => 4
      case Bomb       => 5
    }
  }

  val DataTypeSerializer: DataSerializer[DataType] =
    DataSerializers.BYTE.transform(DataType.fromId(_).get, DataType.toId)
  private val DataTypeKey = EntityDataManager.createKey(classOf[EntityFallingData], DataTypeSerializer)

  implicit val info: EntityInfo[EntityFallingData] = new EntityInfo[EntityFallingData] {
    override def name:                 String            = LibEntityName.FALLING_DATA
    override def create(world: World): EntityFallingData = new EntityFallingData(world)
    override def tracking:             TrackingInfo      = TrackingInfo(range = 40, updateFrequency = 3)
  }
}
class EntityFallingData(
    _world: World,
    @LogicalSideOnly(Side.SERVER) _dataType: DataType,
    @LogicalSideOnly(Side.SERVER) pos: Vector3,
    @LogicalSideOnly(Side.SERVER) var direction: Vector3,
    @LogicalSideOnly(Side.SERVER) var target: Option[Entity],
    @LogicalSideOnly(Side.SERVER) var amount: Float
) extends Entity(_world) {
  if (pos != null && direction != null) {
    setPositionAndRotation(pos.x, pos.y, pos.z, direction.yaw.toFloat, direction.pitch.toFloat)
  }
  if (_dataType != null) {
    dataType = _dataType
  }

  setSize(0.5F, 0.5F)

  def this(world: World) = this(world, null, null, null, None, 0F)

  override protected def entityInit(): Unit =
    dataManager.register(EntityFallingData.DataTypeKey, EntityFallingData.DataType.ScoreGreen)

  override def onUpdate(): Unit = {
    super.onUpdate()
    if (!world.isRemote) {
      val motion = target.fold(direction.multiply(0.25))(Vector3.directionToEntity(this, _))
      motionX = motion.x
      motionY = motion.y
      motionZ = motion.z

      if (!world.isAirBlock(new BlockPos(posX + motionX, posY + motionY, posZ + motionZ))) {
        setDead()
        return
      }

      setPosition(posX + motionX, posY + motionY, posZ + motionZ)
    }
  }

  override def onCollideWithPlayer(player: EntityPlayer): Unit = if (!world.isRemote) {
    dataType match {
      case DataType.ScoreGreen | DataType.ScoreBlue =>
        TouhouHelper.changeAndSyncPlayerData(_.addScore(amount.toInt), player)
        player.playSound(LibSounds.SCORE, 1F, 1F)
      case DataType.Power | DataType.BigPower =>
        TouhouHelper.changeAndSyncPlayerData(_.addPower(amount), player)
      case DataType.Life =>
        TouhouHelper.changeAndSyncPlayerData(_.addLives(amount.toInt), player)
        //player.playSound(LibSounds.EXTEND, 1F, 1F) //TODO: Extend
      case DataType.Bomb =>
        TouhouHelper.changeAndSyncPlayerData(_.addBombs(amount.toInt), player)
    }
    setDead()
  }

  def dataType: DataType = dataManager.get(EntityFallingData.DataTypeKey)

  def dataType_=(dataType: DataType): Unit = dataManager.set(EntityFallingData.DataTypeKey, dataType)

  override protected def readEntityFromNBT(compound: NBTTagCompound): Unit = {
    val targetUuid = compound.getUniqueId("target")
    target = if (targetUuid != null) {
      val targetEntity = world.getPlayerEntityByUUID(targetUuid)
      if (targetEntity == null) NBTHelper.getEntityByUUID(targetUuid, world).toOption else Some(targetEntity)
    } else None
    direction = NBTHelper.getVector(compound, "direction")
    amount = compound.getFloat("amount")
    dataType =
      EntityFallingData.DataType.fromId(compound.getByte("dataType")).getOrElse(EntityFallingData.DataType.ScoreGreen)
  }

  override protected def writeEntityToNBT(compound: NBTTagCompound): Unit = {
    target.foreach(t => compound.setUniqueId("target", t.getUniqueID))
    NBTHelper.setVector(compound, "direction", direction)
    compound.setFloat("amount", amount)
    compound.setByte("dataType", EntityFallingData.DataType.toId(dataType))
  }
}
